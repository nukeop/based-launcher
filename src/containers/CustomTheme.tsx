import { useFlags } from "../hooks/useFlags";

export const CustomTheme = () => {
  const { flags } = useFlags();

  return flags?.theme ? (
    <link rel="stylesheet" type="text/css" href={flags.theme} />
  ) : null;
};
