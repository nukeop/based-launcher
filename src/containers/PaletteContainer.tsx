import { IpcEvent } from "../../common/ipc";
import { Palette } from "../components/Palette/Palette";
import { ArgsContext } from "../contexts/argsContext";
import { ipcRenderer } from "electron";
import Fuse from "fuse.js";
import { useCallback, useContext, useEffect, useState } from "react";

export const PaletteContainer: React.FC = () => {
  const { stdinArgs } = useContext(ArgsContext);
  const options =
    stdinArgs?.map((arg, index) => ({
      id: arg,
      name: arg,
      icon: null,
      onAction: () => onAction(arg),
    })) ?? [];

  const onAction = useCallback((item: string) => {
    ipcRenderer.send(IpcEvent.ReturnSelectedItem, item);
  }, []);

  const [filterInput, setFilterInput] = useState("");
  const [selectedItemIndex, setSelectedItemIndex] = useState(0);

  const fuse = new Fuse(options, {
    keys: ["name", "description"],
    threshold: 0.3,
  });

  const filteredOptions = filterInput
    ? fuse.search(filterInput).map((result) => result.item)
    : options;

  const onUp = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowUp") {
        setSelectedItemIndex(
          selectedItemIndex === 0
            ? filteredOptions.length - 1
            : selectedItemIndex - 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowDown") {
        setSelectedItemIndex(
          selectedItemIndex === filteredOptions.length - 1
            ? 0
            : selectedItemIndex + 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onEnter = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "Enter") {
        filteredOptions[selectedItemIndex].onAction?.();
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  useEffect(() => {
    addEventListener("keydown", onUp);
    addEventListener("keydown", onDown);
    addEventListener("keydown", onEnter);
    return () => {
      removeEventListener("keydown", onUp);
      removeEventListener("keydown", onDown);
      removeEventListener("keydown", onEnter);
    };
  }, [selectedItemIndex, filteredOptions]);

  return (
    <Palette
      options={filteredOptions}
      filterInputValue={filterInput}
      onFilterInputValueChange={setFilterInput}
      selectedItemIndex={selectedItemIndex}
      onSetSelectedItemIndex={setSelectedItemIndex}
    />
  );
};
