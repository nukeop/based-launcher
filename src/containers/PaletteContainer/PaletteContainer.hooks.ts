import { IpcEvent } from "../../../common/ipc";
import { ArgsContext } from "../../contexts/argsContext";
import { useDesktopEntries } from "../../hooks/useDesktopEntries";
import { useFlags } from "../../hooks/useFlags";
import { ipcRenderer } from "electron";
import Fuse from "fuse.js";
import { useCallback, useContext, useEffect, useRef, useState } from "react";
import { VariableSizeList } from "react-window";

export const usePaletteContainerProps = () => {
  const listRef = useRef<VariableSizeList>(null);
  const { stdinArgs, cliFlags, desktopEntries } = useContext(ArgsContext);

  const options =
    stdinArgs?.map((arg, index) => ({
      id: arg,
      name: arg,
      icon: null,
      onAction: () => onAction(arg),
    })) ?? [];

  // const options = desktopEntries?.map((entry) => ({
  //   id: entry["Desktop Entry"].Name,
  //   name: entry["Desktop Entry"].Name,
  //   description: entry["Desktop Entry"].Comment,
  //   icon: entry["Desktop Entry"].Icon,
  //   onAction: () => onAction(entry["Desktop Entry"].Exec),
  // }));

  const onAction = useCallback((item: string) => {
    ipcRenderer.send(IpcEvent.ReturnSelectedItem, item);
  }, []);

  const [filterInput, setFilterInput] = useState("");
  const [selectedItemIndex, setSelectedItemIndex] = useState(0);

  const fuse = new Fuse(options, {
    keys: ["name", "description"],
    threshold: 0.3,
  });

  const filteredOptions = filterInput
    ? fuse.search(filterInput).map((result) => result.item)
    : options;

  const onUp = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowUp") {
        setSelectedItemIndex(
          selectedItemIndex === 0
            ? filteredOptions.length - 1
            : selectedItemIndex - 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onDown = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "ArrowDown") {
        setSelectedItemIndex(
          selectedItemIndex === filteredOptions.length - 1
            ? 0
            : selectedItemIndex + 1
        );
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  const onEnter = useCallback(
    (event: KeyboardEvent) => {
      if (event.key === "Enter") {
        filteredOptions[selectedItemIndex].onAction?.();
      }
    },
    [filteredOptions, selectedItemIndex]
  );

  useEffect(() => {
    listRef.current?.scrollToItem(selectedItemIndex);
  }, [selectedItemIndex]);

  useEffect(() => {
    addEventListener("keydown", onUp);
    addEventListener("keydown", onDown);
    addEventListener("keydown", onEnter);
    return () => {
      removeEventListener("keydown", onUp);
      removeEventListener("keydown", onDown);
      removeEventListener("keydown", onEnter);
    };
  }, [selectedItemIndex, filteredOptions]);

  return {
    listRef,
    flags: cliFlags,
    filteredOptions,
    filterInput,
    setFilterInput,
    selectedItemIndex,
    setSelectedItemIndex,
  };
};
