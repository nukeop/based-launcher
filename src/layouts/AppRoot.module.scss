div.app-root {
    @apply relative w-full h-full border-r-4;
    @apply bg-slate-800/50;
}
