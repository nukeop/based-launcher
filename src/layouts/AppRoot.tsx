import styles from "./AppRoot.module.scss";
import cx from "classnames";

type AppRootProps = {
  className?: string;
  children: React.ReactNode;
};

export const AppRoot: React.FC<AppRootProps> = ({ children, className }) => (
  <div className={cx(styles["app-root"], className)}>{children}</div>
);
