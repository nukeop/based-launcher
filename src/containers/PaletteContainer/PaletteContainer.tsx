import { Palette } from "../../components/Palette/Palette";
import { usePaletteContainerProps } from "./PaletteContainer.hooks";

export const PaletteContainer: React.FC = () => {
  const {
    listRef,
    flags,
    options,
    filterInput,
    setFilterInput,
    selectedItemIndex,
    setSelectedItemIndex,
  } = usePaletteContainerProps();

  return (
    <Palette
      options={options}
      filterInputValue={filterInput}
      onFilterInputValueChange={setFilterInput}
      selectedItemIndex={selectedItemIndex}
      onSetSelectedItemIndex={setSelectedItemIndex}
      listRef={listRef}
      prefixLabel={flags?.inputPrefix}
    />
  );
};
