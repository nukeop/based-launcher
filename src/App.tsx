import { CustomTheme } from "./containers/CustomTheme";
import { PaletteContainer } from "./containers/PaletteContainer/PaletteContainer";
import { ArgsContext } from "./contexts/argsContext";
import { useArgs } from "./hooks/useArgs";
import { useDesktopEntries } from "./hooks/useDesktopEntries";
import { useFlags } from "./hooks/useFlags";
import { AppRoot } from "./layouts/AppRoot";

const App: React.FC = () => {
  const { stdinArgs } = useArgs();
  const { flags } = useFlags();
  const { desktopEntries } = useDesktopEntries();

  return (
    <ArgsContext.Provider
      value={{ stdinArgs, cliFlags: flags, desktopEntries }}
    >
      <AppRoot>
        <CustomTheme />
        <PaletteContainer />
      </AppRoot>
    </ArgsContext.Provider>
  );
};

export default App;
