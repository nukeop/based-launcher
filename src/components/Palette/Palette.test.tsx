import { ArgsContext } from "../../App";
import { PaletteContainer } from "../../containers/PaletteContainer";
import { AppRoot } from "../../layouts/AppRoot";
import {
  fireEvent,
  render,
  RenderResult,
  waitFor,
} from "@testing-library/react";
import { describe, expect, it } from "vitest";

describe("Palette", () => {
  it("renders a palette from piped arguments", () => {
    const component = mountComponent(["first", "second", "third"]);

    expect(component.asFragment()).toMatchSnapshot();
  });

  it("filters entries based on entered text", async () => {
    const component = mountComponent(["first", "second", "third"]);

    const input = await (await component.findAllByTestId("filter-input")).at(0);
    if (input) {
      fireEvent.change(input, { target: { value: "fir" } });
    }

    waitFor(() => expect(component.getByText("first")).toBeTruthy());
    waitFor(() => expect(component.getByText("second")).toBeNull());

    component.unmount();
  });

  it("selects the first item by default", async () => {
    const component = mountComponent(["first", "second", "third"]);

    const firstItem = await nthItem(component, "first");
    expect(firstItem.getAttribute("data-selected")).toBe("true");
  });

  it("right arrow selects the next item", async () => {
    const component = mountComponent(["first", "second", "third"]);
    goRight(component);

    const secondItem = await nthItem(component, "second");
    expect(secondItem.getAttribute("data-selected")).toBe("true");
  });

  it("right arrow wraparound", async () => {
    const component = mountComponent(["first", "second", "third", "fourth"]);
    goRight(component);
    goRight(component);
    goRight(component);

    const fourthItem = await nthItem(component, "fourth");
    expect(fourthItem.getAttribute("data-selected")).toBe("true");
  });

  it("right arrow wraparound in the last row", async () => {
    const component = mountComponent(["first", "second", "third", "fourth"]);
    goRight(component);
    goRight(component);
    goRight(component);
    goRight(component);

    const firstItem = await nthItem(component, "first");
    expect(firstItem.getAttribute("data-selected")).toBe("true");
  });

  it("left arrow selects the previous item", async () => {
    const component = mountComponent(["first", "second", "third"]);
    goRight(component);

    const secondItem = await nthItem(component, "second");
    expect(secondItem.getAttribute("data-selected")).toBe("true");

    goLeft(component);
    const firstItem = await nthItem(component, "first");
    expect(firstItem.getAttribute("data-selected")).toBe("true");
  });

  it("down arrow selects the item in the next row", async () => {
    const component = mountComponent(["first", "second", "third", "fourth"]);
    goDown(component);

    const fourthItem = await nthItem(component, "fourth");
    expect(fourthItem.getAttribute("data-selected")).toBe("true");
  });

  it("up arrow selects the item in the previous row", async () => {
    const component = mountComponent(["first", "second", "third", "fourth"]);

    goDown(component);
    const fourthItem = await nthItem(component, "fourth");
    expect(fourthItem.getAttribute("data-selected")).toBe("true");

    goUp(component);
    const firstItem = await nthItem(component, "first");
    expect(firstItem.getAttribute("data-selected")).toBe("true");
  });

  const mountComponent = (stdinArgs: string[] = []) => {
    return render(
      <ArgsContext.Provider value={{ stdinArgs }}>
        <AppRoot>
          <PaletteContainer />
        </AppRoot>
      </ArgsContext.Provider>
    );
  };

  const goRight = (component: RenderResult) =>
    fireEvent.keyDown(component.getByTestId("palette"), { key: "ArrowRight" });

  const goLeft = (component: RenderResult) =>
    fireEvent.keyDown(component.getByTestId("palette"), { key: "ArrowLeft" });

  const goUp = (component: RenderResult) =>
    fireEvent.keyDown(component.getByTestId("palette"), { key: "ArrowUp" });

  const goDown = (component: RenderResult) =>
    fireEvent.keyDown(component.getByTestId("palette"), { key: "ArrowDown" });

  const nthItem = (component: RenderResult, n: string) =>
    component.getByTestId(`palette-item-${n}`);
});
