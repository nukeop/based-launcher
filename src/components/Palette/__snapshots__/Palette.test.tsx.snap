// Vitest Snapshot v1

exports[`Palette > renders a palette from piped arguments 1`] = `
<DocumentFragment>
  <div
    class="_app-root_v9n3v_1"
  >
    <div
      class="_palette_66plf_1"
      tabindex="-1"
    >
      <input
        class="_input_li5gy_1"
        data-testid="filter-input"
        id="filter-input"
        placeholder="Search..."
        value=""
      />
      <div
        class="_palette-items_66plf_11"
      >
        <div
          class="_palette-grid_1vnjn_1 _columns-3_1vnjn_11"
        >
          <button
            class="_palette-item_40xso_1 _selected_40xso_16"
            data-selected="true"
            data-testid="palette-item-first"
          >
            <div
              class="_palette-item-right_40xso_24"
            >
              <p>
                first
              </p>
              <p
                class="_palette-item-description_40xso_27"
              />
            </div>
          </button>
          <button
            class="_palette-item_40xso_1"
            data-selected="false"
            data-testid="palette-item-second"
          >
            <div
              class="_palette-item-right_40xso_24"
            >
              <p>
                second
              </p>
              <p
                class="_palette-item-description_40xso_27"
              />
            </div>
          </button>
          <button
            class="_palette-item_40xso_1"
            data-selected="false"
            data-testid="palette-item-third"
          >
            <div
              class="_palette-item-right_40xso_24"
            >
              <p>
                third
              </p>
              <p
                class="_palette-item-description_40xso_27"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
